adsws/api/discoverer/solr_route.py

```diff
@@ -0,0 +1,109 @@
+    """
+    try:
+        solr_route = storage.get(solr_route_redis_prefix+user_token)
+    except:
+        solr_route = None
+    return solr_route
+
+def _set_solr_route(storage, solr_route_redis_prefix, user_token, solr_route, solr_route_redis_expiration_time):
+    """
+    Sets the solr route from redis for a given user token. It piggybacks the
+    existing rate limiter extension connection, and if it fails the request
+    will not be stopped (answering a request has a higher priority than
+    assigning a solr instance).
+
+    Keys in redis will expire in N seconds to reduce chances of saturation
+    of a particular solr and to automatically clear entries in redis.
+    """
+    try:
+        storage.setex(solr_route_redis_prefix+user_token, solr_route, solr_route_redis_expiration_time)
+    except:
+        pass
+
+def _build_updated_cookies(request, user_token, solr_route, solr_route_cookie_name):
+    """
+    Based on the current request, create updated headers and cookies content
+    attributes.
+    """
+    # Interpret cookie header content
+    cookies_header = Cookie.SimpleCookie()
+    currrent_cookie_content = request.headers.get('cookie', None)
+    if currrent_cookie_content:
+        cookies_header.load(currrent_cookie_content.encode("utf8"))
+    # Interpret cookies attribute (immutable dict) as a normal dict
+    if request.cookies:
+        cookies = dict(request.cookies)
+    else:
+        cookies = {}
+    # Update cookie structures
+    if solr_route:
+        # Create/update solr route
+        cookies_header[solr_route_cookie_name] = solr_route.encode("utf8")
+        cookies[solr_route_cookie_name] = solr_route
+    else:
+        # Discard non-registered solr route if it is present
+        cookies_header.pop(solr_route_cookie_name, None)
+        cookies.pop(solr_route_cookie_name, None)
+    # Transform cookies structures into the format that request requires
+    cookies_header_content = cookies_header.output(header="", sep=";")
+    cookies_content = ImmutableTypeConversionDict(cookies)
+    return cookies_header_content, cookies_content
+
+def solr_route(storage, solr_route_cookie_name, solr_route_redis_prefix, solr_route_redis_expiration_time):
+    """
+    Assign a cookie that will be used by solr ingress to send request to
+    a specific solr instance for the same user, maximizing the use of solr
+    cache capabilities.
+
+    The storage should be a redis connection.
+    """
+
+    def real_solr_route_decorator(f):
+        @wraps(f)
+        def decorated_function(*args, **kwargs):
+            # Obtain user token, giving priority to forwarded authorization field (used when a microservice uses its own token)
+            user_token = request.headers.get('X-Forwarded-Authorization', None)
+            if user_token is None:
+                user_token = request.headers.get('Authorization', None)
+            if user_token and len(user_token) > 7: # This should be always true
```

Comment:
occurred to me that this depends on the oauth decorator running first (which is the case), however as it stands right now, it is conditional:

https://github.com/marblestation/adsws/blob/1013364babcb059dd7ec5daaa1ff31fc83ea55e3/adsws/api/discoverer/utils.py#L81

which might be OK (for now); but I have got a question for ya (since you might have been digging inside); when I am authenticated through a session cookie (ie. not sending the token), is the user object (and oauth info) loaded before your sroute gets called?

Reply:
I have done some testing and we cannot actually make a request to solr-service only with a session cookie, it requires the bearer token or we just get "Unauthorized".

Reply:
while that is probably true; there are endpoints, such as /user where an user is granted access with just a cookie; and so if this function expects user_token we should verify that it is there

Reply:
But the solr_route decorator is only for the solr-service, requests to /user do not go through the solr_route decorator. Do you know a microservice that grants access just with a session cookie and that contact solr-service for something?